### PR TITLE
[browser] improve behavior after exit, better docs

### LINKED ISSUE #94906: [browser] improve behavior after exit
We should

- be louder about it when you are calling into stopped runtime
        was there any message in the console ?
        was there unhandled exception from your onClick ?
- the error messages could point out what's the fix
- change the browser template to use exit-less runMain and keep one with exit only for console template
- dotnet.run jsdoc should say that it's going to stop the runtime

### PR DESCRIPTION
**Pull request by pavelsavara**
- exit assert message to log
 - browser template doesn't exit by default
 - improved docs

Fixes https://github.com/dotnet/runtime/issues/94906

### CODE REVIEW

File: src/mono/browser/runtime/dotnet.d.ts [hunk 1]
```diff
@@ -54,27 +54,79 @@ type InstantiateWasmCallBack = (imports: WebAssembly.Imports, successCallback: I
...
+     * level == 0 disables debugging and enables interpreter optimizations
+     * level < 0 enables debugging and disables debug logging.
+     */
     withDebugging(level: number): DotnetHostBuilder;
+    /**
+     * @param mainAssemblyName Sets the name of the assembly with the Main() method. Default is the same as the .csproj name.
+     */
     withMainAssembly(mainAssemblyName: string): DotnetHostBuilder;
+    /**
+     * Capture "command line" arguments for the Main() method from browser query string.
```
Comment (maraf) [inline src/mono/browser/runtime/dotnet.d.ts:101]:
```suggestion
     * Supply "command line" arguments for the Main() method from browser query arguments named "arg". Eg. `index.html?arg=A&arg=B&arg=C`.
```

File: src/mono/browser/runtime/dotnet.d.ts [hunk 2]
```diff
@@ -54,27 +54,79 @@ type InstantiateWasmCallBack = (imports: WebAssembly.Imports, successCallback: I
...
+     * Starts the runtime and returns promise of the API object.
+     */
     create(): Promise<RuntimeAPI>;
-    run(): Promise<number>;
+    /**
+     * Runs the Main() entrypoint of the application and resolves promise with exit code.
+     * @param keepRunning if true, the runtime will keep processing APIs calls after the Main() method resolves. Default is false.
+     * You can provide "command line" arguments for the Main() method using
+     * - dotnet.withApplicationArguments()
+     * - dotnet.withApplicationArgumentsFromQuery()
```
Comment (maraf) [inline src/mono/browser/runtime/dotnet.d.ts:127]:
```suggestion
     * @param keepRunning if true, the runtime will keep processing API calls after the Main() method exits. Default is false.
     * You can provide "command line" arguments for the Main() method using
     * - dotnet.withApplicationArguments(["A", "B", "C"])
     * - dotnet.withApplicationArgumentsFromQuery()
```

File: src/mono/browser/runtime/dotnet.d.ts [hunk 3]
```diff
@@ -373,45 +425,179 @@ type DotnetModuleConfig = {
     exports?: string[];
 } & Partial<EmscriptenModule>;
 type APIType = {
+    /**
+     * Runs the Main() entrypoint of the application and exits the runtime.
```
Comment (maraf) [inline src/mono/browser/runtime/dotnet.d.ts:429]:
```suggestion
     * Runs the Main() method of the application and exits the runtime.
```

File: src/mono/browser/runtime/dotnet.d.ts [hunk 4]
```diff
@@ -54,27 +54,79 @@ type InstantiateWasmCallBack = (imports: WebAssembly.Imports, successCallback: I
...
      * from a custom source, such as an external CDN.
      */
     withResourceLoader(loadBootResource?: LoadBootResourceCallback): DotnetHostBuilder;
+    /**
+     * Starts the runtime and returns promise of the API object.
+     */
     create(): Promise<RuntimeAPI>;
-    run(): Promise<number>;
+    /**
+     * Runs the Main() entrypoint of the application and resolves promise with exit code.
```
Comment (maraf) [inline src/mono/browser/runtime/dotnet.d.ts:123]:
```suggestion
     * Runs the Main() method of the application and resolves promise with exit code.
```

File: src/mono/browser/runtime/dotnet.d.ts [hunk 5]
```diff
@@ -373,45 +425,179 @@ type DotnetModuleConfig = {
...
+    /**
+     * Runs the Main() entrypoint of the application and exits the runtime.
+     * Note: this will keep the dotnet runtime alive and the APIs will be available for further calls.
+     * @param mainAssemblyName name of the assembly with the Main() method.
+     * @param args command line arguments for the Main() method.
+     * @returns exit code of the Main() method.
+     */
     runMain: (mainAssemblyName: string, args?: string[]) => Promise<number>;
+    /**
+     * Runs the Main() entrypoint of the application and exits the runtime.
```
Comment (maraf) [inline src/mono/browser/runtime/dotnet.d.ts:437]:
```suggestion
     * Runs the Main() method of the application and exits the runtime.
```

File: src/mono/browser/runtime/dotnet.d.ts [hunk 6]
```diff
@@ -373,45 +425,179 @@ type DotnetModuleConfig = {
...
+     */
     runMainAndExit: (mainAssemblyName: string, args?: string[]) => Promise<number>;
+    /**
+     * Sets the environment variable for the "process"
+     * @param name
+     * @param value
+     */
     setEnvironmentVariable: (name: string, value: string) => void;
+    /**
+     * Loads the assembly with the given name and returns the [JSExports] of the assembly
```
Comment (maraf) [inline src/mono/browser/runtime/dotnet.d.ts:451]:
```suggestion
     * Returns the [JSExport] methods of the assembly with the given name
```

File: src/mono/browser/runtime/dotnet.d.ts [hunk 7]
```diff
@@ -373,45 +425,179 @@ type DotnetModuleConfig = {
...
+     * Returns the configuration object used to start the runtime.
+     */
     getConfig: () => MonoConfig;
+    /**
+     * Executes scripts which were loaded during runtime bootstrap.
+     * You can register the scripts using MonoConfig.resources.modulesAfterConfigLoaded and MonoConfig.resources.modulesAfterRuntimeReady.
+     */
     invokeLibraryInitializers: (functionName: string, args: any[]) => Promise<void>;
+    /**
+     * write into the WASM linear memory
```
Comment (maraf) [inline src/mono/browser/runtime/dotnet.d.ts:471]:
```suggestion
     * Writes to the WASM linear memory
```

File: src/mono/browser/runtime/dotnet.d.ts [hunk 8]
```diff
@@ -373,45 +425,179 @@ type DotnetModuleConfig = {
...
+     */
     setHeapF32: (offset: NativePointer, value: number) => void;
+    /**
+     * write into the WASM linear memory
+     */
     setHeapF64: (offset: NativePointer, value: number) => void;
+    /**
+     * Read from the WASM linear memory
```
Comment (maraf) [inline src/mono/browser/runtime/dotnet.d.ts:519]:
```suggestion
     * Reads from the WASM linear memory
```

File: src/mono/browser/runtime/dotnet.d.ts [hunk 9]
```diff
@@ -373,45 +425,179 @@ type DotnetModuleConfig = {
...
+     */
     getHeapF32: (offset: NativePointer) => number;
+    /**
+     * Read from the WASM linear memory
+     */
     getHeapF64: (offset: NativePointer) => number;
+    /**
+     * Short term view of the WASM linear memory. Don't store the reference, don't use after await.
```
Comment (maraf) [inline src/mono/browser/runtime/dotnet.d.ts:567]:
```suggestion
     * Returns a short term view of the WASM linear memory. Don't store the reference, don't use it after await.
```

File: src/mono/browser/runtime/loader/exit.ts [hunk 10]
```diff
@@ -13,8 +13,8 @@ export function is_runtime_running() {
 }
 
 export function assert_runtime_running() {
-    mono_assert(runtimeHelpers.runtimeReady, "mono runtime didn't start yet");
-    mono_assert(!loaderHelpers.assertAfterExit || !is_exited(), () => `mono runtime already exited with ${loaderHelpers.exitCode} ${loaderHelpers.exitReason}`);
+    mono_assert(runtimeHelpers.runtimeReady, "dotnet runtime didn't start yet. Please call dotnet.create() first.");
```
Comment (maraf) [inline src/mono/browser/runtime/loader/exit.ts:16]:
```suggestion
    mono_assert(runtimeHelpers.runtimeReady, ".NET runtime didn't start yet. Please call dotnet.create() first.");
```

File: src/mono/browser/runtime/loader/exit.ts [hunk 11]
```diff
@@ -13,8 +13,8 @@ export function is_runtime_running() {
...
+    mono_assert(!loaderHelpers.assertAfterExit || !is_exited(), () => `dotnet runtime already exited with ${loaderHelpers.exitCode} ${loaderHelpers.exitReason}. You can pass dotnet.run(false) or use runtime.runMain() instead.`);
```
Comment (maraf) [inline src/mono/browser/runtime/loader/exit.ts:17]:
```suggestion
    mono_assert(!loaderHelpers.assertAfterExit || !is_exited(), () => `.NET runtime already exited with ${loaderHelpers.exitCode} ${loaderHelpers.exitReason}. You can pass dotnet.run(false) or use runtime.runMain() instead.`);
```

**Review by maraf — COMMENTED** [review]
Some suggestions, not that I'm an expert here

File: src/mono/browser/runtime/dotnet.d.ts [hunk 12]
```diff
@@ -54,27 +54,79 @@ type InstantiateWasmCallBack = (imports: WebAssembly.Imports, successCallback: I
...
     withDiagnosticTracing(enabled: boolean): DotnetHostBuilder;
+    /**
+     * @param level
+     * level > 0 enables debugging and sets the debug log level to debugLevel
+     * level == 0 disables debugging and enables interpreter optimizations
+     * level < 0 enables debugging and disables debug logging.
+     */
     withDebugging(level: number): DotnetHostBuilder;
+    /**
+     * @param mainAssemblyName Sets the name of the assembly with the Main() method. Default is the same as the .csproj name.
```
[the remark this answers is not in the document]
Comment (pavelsavara) [reply src/mono/browser/runtime/dotnet.d.ts:97]: this is good enough for most people.

File: src/mono/browser/runtime/types/index.ts [hunk 13]
```diff
@@ -396,7 +581,13 @@ export type RuntimeAPI = {
 } & APIType
 
 export type ModuleAPI = {
+    /**
+     * The builder for the dotnet runtime.
```
Comment (maraf) [inline src/mono/browser/runtime/types/index.ts:585]:
```suggestion
     * The builder for the .NET runtime.
```

**Review by maraf — APPROVED** [review]

**Pull request merged**